EXP: Backend Resource (#4488)
kubernetes-sigs/gateway-api: geps/gep-4488/index.md

Reviewer: Mostly just thinking out loud: if Session Persistence is a candidate for Backend, we need to determine if Session Persistence is really a consumer-side configuration. 

Two questions help clarify the producer vs consumer fit (other than the "who"): 

1. **Can it be fragmented per-consumer?** Yes, each consumer's session pinning can be independent. Cookie name collisions are possible, but that's true with producer-side (`BackendTrafficPolicy`) config too.
2. **Is app statefulness intrinsic to the service?** Maybe. Most modern apps are flexible with state. This argument carries less weight than it used to.

So I think Session Persistence can be arguably both producer and consumer, which means it could fit into Backend. I can follow up with this with another discussion about Backend attachment here https://github.com/kubernetes-sigs/gateway-api/discussions/4462. 

There's also an interesting argument about per-path persistence: the counter point to route-inline for using BackendTrafficPolicy has always been "well just split up your services," but that felt heavy because it meant creating duplicate Services with all that overhead. Backend makes this argument even stronger: it's more lightweight now. Two Backends pointing to the same endpoints with different configs, no duplicate Services needed.

Author: I need to make some time to go deeper into that discussion, but I think the semantic clarification I added at the beginning of the GEP will help here. If session persistence is going to change the characteristics of the outbound HTTP request (and I believe it does), then it definitely is a candidate for Backend. The signaling mechanism for what key etc. is a different problem to be solved in a different place

Reviewer: Would health checks be a consumer-side configuration?

Author: For active health checking, yteah probably

Author: We should consider advising implementations to offer (require?) a list of allowed domains in the control plane. Similar to the mitigation of https://github.com/zalando/skipper/security/advisories/GHSA-mxxc-p822-2hx9

Author: Added a SHOULD in the GEP for this

# GEP-4488: Backend Resource for Egress and Service Decoration

* Issue: [#4488](https://github.com/kubernetes-sigs/gateway-api/pull/4488)
  * Incubated by the [AI Gateway Working Group](https://github.com/kubernetes-sigs/wg-ai-gateway/pull/20)
* Status: Provisional

## TLDR

> This GEP proposes a new `Backend` resource to address limitations in external destination representation, policy application, and security concerns with the current Service-based backend system. The namespace-scoped `Backend` resource serves as a consumer-focused alternative to synthetic Services for external hosts and provides a foundation for service decoration and enhanced policy application.

Reviewer 2: I think that we actually should go broader here.

If we're going to have a general `Backend` resource in Gateway API, then that should be the place where we do a lot of things that we can't do with Service,  like:

* hostname/ExternalName support
* TLS Settings
* Session Persistence settings
* Maybe even circuit breaking settings

Basically, anything that we _wish_ we could put into Service can go here instead, with egress being the _first_ use case, but not the last.

It doesn't change the scope of this first iteration that much, I think that we should just be clear that:

* Backend is effectively a Gateway-specific redesign of Service, which allows to add more things
* The first thing to add is a safer version of ExternalName, which will be protected by an Extended feature.

That leaves space for us to add other use cases as further Extended features, with Core conformance for Backend just covering the same functionality as Service is now.

How about something like this?

```suggestion
This GEP proposes a new `Backend` resource to address the fact that Service is a mature and stable, if complex API. It is useful to be able to add Gateway-specific behaviors to Service, but previous approaches (like BackendTLSPolicy) have significant limltations, particuarly around discoverability and implementation complexity.

In particular, AI use cases have an urgent need to be able to better handle external destination representation, policy application, and security concerns than the current Service-based backend system can manage.

The namespace-scoped `Backend` resource serves in this use case as a consumer-focused alternative to synthetic Services for external hosts and provides a foundation for service decoration and enhanced policy application, as a first use for this new resources.

Further use of the `Backend` resource for other configuration that is tightly bound to Service is expected to follow.
```

Author: Yeah I started with the smaller scope first since egress was the immediate need, but I'm totally aligned with the direction you've laid out. I think the suggestion makes that direction clear without explicitly committing too much (which was my only concern)

The Backend resource is explicitly designed as a **consumer resource** - it describes how a gateway should connect to a destination from the client perspective, regardless of whether that destination is internal or external to the cluster.

## API Specification

### Backend Resource Schema

```go
type Backend struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`
  Spec   BackendSpec   `json:"spec"`
  Status BackendControllerStatus `json:"status,omitempty"`
}

// +kubebuilder:validation:Enum=Hostname;EndpointSelector
type BackendType string

const (
  BackendTypeHostname             BackendType = "Hostname"
  BackendTypeEndpointSelector     BackendType = "EndpointSelector"
)

// +kubebuilder:validation:ExactlyOneOf=Hostname,EndpointSelector
type BackendSpec struct {
  // Type defines the destination type
  Type BackendType `json:"type"`

  // Ports defines the destination ports and protocols
  // kubebuilder:validation:MinItems=1
  // kubebuilder:validation:MaxItems=16
  Ports []BackendPort `json:"ports,omitempty"`

  // Hostname specifies the configuration for an Hostname backend. Only used if type is Hostname.
  // +optional

>   Hostname *HostnameBackend `json:"hostname,omitempty"`

Reviewer 2: I think this should be guarded behind an Extended feature. That way, implementations can choose whether or not they want to support this, accepting the security risks, and it's up to them to inform users about it.

I anticipate that AI use cases will all support the Extended feature anyway, and so it shouldn't add too much overhead. But an Extended feature will allow us to handle the "this field is not supported" use case much more clearly.

Author: I'm ok with Hostname being Extended initially, I just want to be careful not to conflate egress with AI. Egress is plenty useful on its own without AI and has been a missing part of k8s networking for a while

Author: Moved hostname to extended

  // conditions represent the current state of the Backend resource.
  // Each condition has a unique type and reflects the status of a specific aspect of the resource.
  //
  // Standard condition types include:
  // - "Available": the resource is fully functional
  // - "Progressing": the resource is being created or updated
  // - "Degraded": the resource failed to reach or maintain its desired state
  //
  // The status of each condition is one of True, False, or Unknown.
  // +listType=map
  // +listMapKey=type
  // +optional
  Conditions []metav1.Condition `json:"conditions,omitempty"`
}

type BackendPort struct {
  // Number defines the port number of the backend.
  // +required
  // +kubebuilder:validation:Minimum=1
  // +kubebuilder:validation:Maximum=65535
  Number uint32 `json:"number"`
  // Protocol defines the protocol of the backend.
  // +optional
  // +kubebuilder:validation:MaxLength=256
  Protocol BackendProtocol `json:"protocol"`
  // TLS defines the TLS configuration that a client should use when talking to the backend.
  // TODO: To prevent duplication on the part of the user, maybe this should be declared once at the

>   // top level with per-port overrides?

Reviewer 2: When it's all in the one resource like this, I think that more explicit should be our default. A default with per-port overrides adds a _lot_ of cognitive complexity when compared to putting the TLS settings into this struct, especially  when it's likely that a lot of this config will be generated anyway.

Author: I tend to agree; that's why I went with this initial implementation, but left the comment here for more discussion

Author: Resolving in favor of the other thread with John

  // +optional
  TLS *BackendTLS `json:"tls,omitempty"`
}
```

### Hostname Backend Configuration

```go
type HostnameBackend struct {
  // Address specifies the destination address used to reach this hostname.

>   Address string `json:"address"`

Reviewer 2: If this is a `hostname`, should it use the existing Hostname type? That way, there's some validation? Also, we should clarify if IP addresses are valid here (both for IPv4 and IPv6). I strongly recommend not allowing that, as it will make the security concerns even harder to manage than they already are.

Author: Good call; moved to PreciseHostname

     - [GEP 3875](https://github.com/kubernetes-sigs/gateway-api/pull/3876) proposed, among other things, adding consumer overrides to `BackendTLSPolicy`; however, that proposal introduced several new fields to the resource, including a `from` selector on `targetRef` that would have added significant complexity to both the API and implementations. Furthermore, the GEP has a [limitation](https://github.com/kubernetes-sigs/gateway-api/pull/3876/changes#diff-67a0076fb272af6273ce353d4687732735e03ddec8ae2bbc35b0a905281f9057R88) that it would not be possible to enforce consumer-side policies originating from the same namespace as the producer. This proposal was ultimately abandoned.

3. **Per-Backend Client Certificates**
   - External destinations may require different client certificates
   - Current Gateway API patterns only support one client certificate per Gateway
   - Backend-specific client certificates are essential for many external integrations

#### Arguments for Policy-Based Configuration

1. **API Consistency**
   - Gateway API uses policy attachment for most configuration beyond basic routing
   - Inline configuration creates another place to define TLS settings

>    - Multiple places for TLS configuration could lead to misalignment of features

Reviewer 2: All of these things are true, but they are only true because _we can't add anything to Service_. If we could we would have done so; a separate resource like this is a much better idea, that I wish we had moved ahead with originally.

BackendTLSPolicy is around now because we made it Stable, but I don't think that is a good reason to continue pushing forward with _more_ Policy config.

Author: Added a note to this effect

**Cluster-Admin Risk Acceptance**
- Cluster administrators who grant Backend creation permissions accept DNS trust model
- Network-level controls (firewalls, proxy configuration) provide defense in depth
- Backend resources provide audit trail for external dependencies

## EndpointSelector Type

TODO: Reference GEP 4731 once it merges.

## Extension Framework

The Backend resource provides two levels for applying extensions and policies:

### 1. Route-Level Extensions (HTTPRoute Filters)

Applied to individual requests as they are routed to a Backend.

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
spec:
  rules:
  - matches:
    - path: { value: "/api/models" }
    filters:
    - type: ExtensionRef
      extensionRef:
        name: rate-limiter
        kind: RateLimitPolicy
    backendRefs:
    - name: openai-api
      kind: Backend
```

### 2. Policy Attachment

Separate policy resources attached to Backend resources.

```yaml
apiVersion: networking.example.com/v1
kind: RetryPolicy
metadata:
  name: openai-retry
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Backend
    name: openai-api
  retries: 3

>   backoff: exponential

Reviewer 2: This is fine for ImplementationSpecific things, but I would argue that if we do this in upstream, we should addi them into Backend instead. It's easier to understand, and very useful for other cases as well.

Reviewer 3: I don't disagree with this but curious what this means. 

Lets say we make "retry" feature (ignore that it exists today on route :slightly_smiling_face: ). We decide to put it on Backend instead of as a policy. A user has an existing HTTPRoute pointing to a Service.

How do they consume the retry feature? do they need to migrate from Service to Backend + EndpointSelector?

Author: I would say yes they need to migrate, but over the long-term, that's not going to be a huge issue since `Service` will create `EndpointSelector` resources that a `Backend` can bind to.

## Proposal

The Backend resource addresses a fundamental gap in Gateway API's backend reference system: first-class support for external destinations. Currently, external APIs, databases, and services must be represented as synthetic Kubernetes Services with `type: ExternalName`, which creates security vulnerabilities, policy application challenges, and resource management overhead. Furthermore, `Backend`s of type `EndpointSelector` can be used to decorate existing `Service`s (or other resources that fulfill the backend role) with TLS and protocol configuration.

The Backend resource is explicitly designed as a **consumer resource** - it describes how a gateway should connect to a destination from the client perspective, regardless of whether that destination is internal or external to the cluster.

## API Specification

### Backend Resource Schema

```go
type Backend struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`
  Spec   BackendSpec   `json:"spec"`
  Status BackendControllerStatus `json:"status,omitempty"`
}

type BackendSpec struct {
  // Destination defines where traffic should be sent

>   Destination BackendDestination `json:"destination"`

Reviewer 3: nit: whats the point of a top level field everything is nested under? I would assume the reason is future-proofing but I cannot think of anything that would not be under Destination?

Author: I think the reason was to make the fields within it a oneOf. Kubebuilder may not require that though (I think in protobuf b/c istio)

Reviewer 3: Yeah I cannot think of any reason that would be needed. kubebuilder can do it no problem

Author: Fixed

}

type BackendType string

const (
  BackendTypeHostname             BackendType = "Hostname"
  BackendTypeEndpointSelector BackendType = "EndpointSelector"
)

type BackendDestination struct {
  // Type defines the destination type
  Type BackendType `json:"type"`

  // Ports defines the destination ports and protocols
  // kubebuilder:validation:MinItems=1
  // kubebuilder:validation:MaxItems=16

>   Ports []BackendPort `json:"ports,omitempty"`

Reviewer 3: Sorry this is a very late comment so I imagine its already been discussed. But IMO a singular port is best.

This avoids awkward nesting of policies under `.ports`, the need to specify a port on backendRefs, etc.

Also, the majority of the time you only have 1 port. And if you have 2 ports, its because they are meaningfully different -- different policies (especially tls!) etc.

Author: > Also, the majority of the time you only have 1 port. And if you have 2 ports, its because they are meaningfully different -- different policies (especially tls!) etc.

Sure but wouldn't control planes need to do merging across different Backends to the same service but different ports anyway? And ensure no duplicates, etc. Having per port stuff with section name references matches how users think about backends IMHO

Reviewer 3: Why would they need to? Just thinking of Envoy, the translation is to a Cluster which is going to be 1 per-port. So while my motivations are not at all for making it easier for the implementation, it does seem easier (1:1 vs 1:N)?

It may be slightly easier for the Backend creator but it pushes the problem to the route authors IMO

Author: That's fair...I do think for the EndpointSelector case, multiple ports is nicer because you avoid having N EndpointSlices for N ports. By grouping them in 1 resource, you have the ability to consider all the ports as part of one resource if you need to but if we make it 1 port per resource, we lose that ability

Reviewer 3: Wouldn't the EndpointSelector make the EndpointSlice and then the Backend just references it? So if you want 2 ports you have 2 Backend, 1 EndpointSelector, 1 EndpointSlice

Author: The current draft of the endpoint selector GEP (https://github.com/kubernetes-sigs/gateway-api/pull/4731) has each backend of type EndpointSelector actually create the EndpointSelector resource in the controller. That UX isn't 100% final yet but that's what I was working around

Reviewer 2: I agree with @howardjohn on this one, multiple ports adds at least an order of magnitude to the overall complexity of the implementation. I would actually recommend starting with this:

I agree with John that I think it will be unusual that multiple ports on a backend will share the exact same set of security policies, TLS settings, load balancing algorithms, etc. That's relevant, because if they don't share the exact same set of all of those things, then you need to create separate Backends to describe them anyway.

So, my suggested solution for now is that we have a slice of backendPorts, where the only supported value to begin with is 1. (we do this on the CA certificate reference on BackendTLSPolicy as well). Implementations _may_ do their own thing with it, but the standard behavior is one.

Author: I hadn't considered more than just TLS, but even if I think about the policy attachment aspect, a sectionName would end up being necessary to be able to target a specific port anyway! So I think you've both convinced me; I like the idea of keeping the type as a slice, but restricting MaxLength to 1.

Reviewer 3: > So, my suggested solution for now is that we have a slice of backendPorts, where the only supported value to begin with is 1. (we do this on the CA certificate reference on BackendTLSPolicy as well). Implementations may do their own thing with it, but the standard behavior is one.

FWIW if we do this we do lose one benefit of just having 1: that we need `spec.ports.tls` instead just `spec.tls`.

Not a bit deal and probably worth it if we are ever going to do multiple ports, but a slight UX hit.

(We could also have top level and per-port overrides, should we want to start with `spec.tls` now and expand in the future to `spec.ports.tls` where port field overrides the top level field)

Author: Hmm and to my earlier point about any policy attachment (in-project or implementation specific), sectionName would be required anyway...the decision at this point isn't a huge issue since this is still experimental, so I'm inclined to shift to 1 Backend resource per port, experiment with that, and see if something makes us really want to combine ports within a backend.

> (We could also have top level and per-port overrides, should we want to start with spec.tls now and expand in the future to spec.ports.tls where port field overrides the top level field)

This would still remain an option despite the UX concerns [here](https://github.com/kubernetes-sigs/gateway-api/pull/4488#discussion_r3151920759)

Author: Made the move to explicitly having 1 port

}

type BackendPort struct {
  // Number defines the port number of the backend.
  // +required
  // +kubebuilder:validation:Minimum=1
  // +kubebuilder:validation:Maximum=65535
  Number uint32 `json:"number"`
  // Protocol defines the protocol of the backend.
  // +required
  // +kubebuilder:validation:MaxLength=256
  Protocol BackendProtocol `json:"protocol"`
  // TLS defines the TLS configuration that a client should use when talking to the backend.
  // TODO: To prevent duplication on the part of the user, maybe this should be declared once at the
  // top level with per-port overrides?
  // +optional
  TLS *BackendTLS `json:"tls,omitempty"`
}
```

### Hostname Backend Configuration

```go
type HostnameBackend struct {
  // Address specifies the destination address used to reach this hostname.
  Address string `json:"address"`
}
```

### Protocol and Extension Support

```go

// BackendProtocol defines the protocol for backend communication.
// +kubebuilder:validation:Enum=HTTP;HTTP2;TCP;MCP
type BackendProtocol string

const (
  BackendProtocolHTTP  BackendProtocol = "HTTP"
  BackendProtocolHTTP2 BackendProtocol = "HTTP2"

>   BackendProtocolTCP   BackendProtocol = "TCP"

Reviewer 3: What does it mean for a user to specify TCP vs HTTP when traffic is coming from an HTTPRoute (i.e. the traffic is already HTTP)?

Author: Good call out; these are really only necessary when addressing frontends so no need for them here

Author: Reduced to just MCP with comments about this being for higher-level protocols

>   BackendProtocolMCP   BackendProtocol = "MCP"

Reviewer 3: Similar to TCP; what does it actually mean from a user perspective to declare this TCP vs HTTP vs MCP? All of them are "true" for an MCP backend. Is it impacting the behavior?

Author: As above, you're right; the route already tells the implementation the L7 protocol. Now if there are higher level protocols needed (e.g.MCP, LLM, etc.), then I think this becomes more useful

Reviewer 3: Does the route? For MCP today you would use HTTPRoute

Author: That's what I'm saying in the second sentence; for higher level protocols like MCP or things like that, BackendProtocol becomes more useful

Author: Reduced to just MCP with comments about this being for higher-level protocols

Reviewer 2: I know that AI use cases are the primary driver here, but it's important that we ensure this is useful even in the case it's _not_ used for AI.

So, in the case that Backend is used for something that's _not_ AI, what could this be set to? If we're using Backend solely as a BackendTLSPolicy replacement, what would this be set to? Would it be unset, or would it be set to "TLS"? If the communication is in the clear, what should this be set to?

### Tradeoffs: Inline TLS vs. Policy-Based TLS

#### Arguments for Inline TLS Configuration

1. **Simplified UX for External Destinations**
   - External FQDNs often require TLS configuration that is specific to that destination
   - Better discoverability for users who want to understand what TLS settings apply for backends within a route
   - Much simpler for implementations to integrate due to TLS settings being colocated with destination

2. **BackendTLSPolicy Limitations**
   - Current `BackendTLSPolicy` is designed around Service-based backends only
   - `BackendTLSPolicy` currently does not support per-consumer overrides
   - It is unclear whether `BackendTLSPolicy` is a producer or consumer oriented resource

3. **Per-Backend Client Certificates**
   - External destinations may require different client certificates
   - Current Gateway API patterns only support one client certificate per Gateway
   - Backend-specific client certificates are essential for many external integrations

#### Arguments for Policy-Based Configuration

> 1. **API Consistency**

Reviewer 3: No need to add into the doc, but I would also add that it puts third-party policies and first-party ones on even footing.

Reviewer 2: I think this is a fair point, but I also think that inlining then makes conflict resolution easier, as inline is the tightest possible scope, that can win conflict resolution by default.

3. **Type Definition Alignment**: Align the inline TLS types with `BackendTLSPolicy` types as closely as possible for consistency

### Implementation Example

```yaml
# Gateway-level TLS remains authoritative for incoming connections
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
spec:
  listeners:
  - name: https
    protocol: HTTPS
    tls:
      certificateRefs:
      - name: gateway-cert
  # Deprecate gateway.spec.tls.backend in favor of Backend resource configuration

---
# Backend resource with inline TLS for external destination
apiVersion: gateway.networking.k8s.io/v1alpha1
kind: Backend
metadata:
  name: openai-api
  namespace: ai-apps
spec:
  destination:
    type: Hostname
    hostname:
      address: api.openai.com
      tls:
        serverName: api.openai.com
        clientCertificate:
          secretRef:
            name: openai-client-cert
            key: tls.crt
    ports:
    - name: https
      port: 443
      protocol: HTTPS

---
# HTTPRoute referencing Backend
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
spec:
  rules:

>   - backendRefs:

Reviewer 3: This needs a port (unless we adopt my other comment recommending making a single port).

This also highlights, IMO, why I think we should do that. As an HTTPRoute author, I don't care what port openai is on, just like I don't care what its hostname is or policies to reach it. I just want to reach it.

Author: Good catch. I don't think port is a huge lift though since we already require it for `Service` right?

Author: Added port

Reviewer 2: Yeah, it's already required, but again, that's because Service allows multiple ports, and so it has to be. If Backend did not allow multiple ports, we could make `port` not required when the Kind is Backend.

Reviewer 2: To put this another way, would you really want to be able to target OpenAI on multiple TLS ports at once? That seems.. not very useful. If the TLS is on multiple ports, it's probably because those TLS connections have different properties, which should mean there are different Backends to describe them. As we add features to Backend (like LB algorithms etc), that problem will only worsen.

# GEP-4488: Backend Resource for Egress and Service Decoration

* Issue: [#4488](https://github.com/kubernetes-sigs/gateway-api/pull/4488)
  * Incubated by the [AI Gateway Working Group](https://github.com/kubernetes-sigs/wg-ai-gateway/pull/20)
* Status: Provisional

## TLDR

> This GEP proposes a new `Backend` resource to address the fact that Service is a mature and stable, if complex API. It is useful to be able to add Gateway-specific behaviors to Service, but previous approaches (like BackendTLSPolicy) have significant limltations, particuarly around discoverability and implementation complexity.

Reviewer 4: The limitations you listed later are only applicable when BackendTLSPolicy is applied to ExternalName services.  It should be made clearer in both the name and purpose of the new resource, that it is primarily needed for cases where ExternalName services were the only available option.

Author: I addressed this by clarifying that the scope of Backend is wider than just egress.

The namespace-scoped `Backend` resource serves in this use case as a consumer-focused alternative to synthetic Services for external hosts and provides a foundation for service decoration and enhanced policy application, as a first use for this new resources.

Further use of the `Backend` resource for other configuration that is tightly bound to Service is expected to follow.

## Motivation

The Kubernetes `Service` resource conflates two distinct concerns that have become increasingly problematic as Gateway API adoption grows:

1. **Frontend concerns**: How services are discovered and called (DNS names, ClusterIPs, service discovery)
2. **Backend concerns**: Where traffic should be routed and how to connect to destinations (endpoints, TLS configuration, protocol settings)

This conflation creates several significant issues:

### External Destination Limitations

Currently, representing external destinations in Gateway API requires synthetic `Service` objects with `type: ExternalName`, which:

- **Security vulnerabilities**: ExternalName Services are subject to DNS rebinding attacks ([CVE-2021-25740](https://github.com/kubernetes/kubernetes/issues/103675))

> - **Policy limitations**: Cannot apply backend-specific policies (TLS, authentication, rate limiting) without affecting all consumers

Reviewer 4: This is by design.  In the case where a backend is a service then a policy should affect all consumers.  There shouldn't be singleton application of policy characteristics.

Author: IMO, that's fundamentally incorrect for every case *except* ingress. In mesh and egress contexts, the producer of the Service shouldn't tell consumers what to do. Ingress is actually the exception and that only works because ingress gateways are the main consumer!

Reviewer 4: Like @youngnick, [who explained it much better than me](https://github.com/kubernetes-sigs/gateway-api/pull/4488#discussion_r3186506002), I have ingress as my main concern.  I can see the point of an egress gateway too, and still don't see why something like an egress gateway can't be used instead of ExternalName.  Mesh is the exception for those of us who don't work with it.

IMO we get into trouble when we try to apply a resource that works for both ingress and mesh.  The complexity grows to accommodate multiple cases, or users get two ways to do something without understanding which to choose.

Looking forward to updates on this.  Would be nice to see a new name that could not be confused with the [definition of backend](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-2907/index.md?plain=1#L49-L50), nor unable to be reconciled with some non-existent "frontend" resource.

Author: I was able to talk with Nick and some others, and I added some clarifying language to the beginning of the GEP. I don't think you're alone in your impression of what a "backend" is, so please let me know if the added context is helpful

- **Synthetic resource overhead**: Creates artificial Kubernetes resources for external dependencies

### Policy Application Complexity

The overloaded Service resource makes policy attachment ambiguous:

- **Producer vs Consumer confusion**: Services represent both "how to call me" (producer) and "where to route" (consumer) concerns
- **Policy scope unclear**: TLS policies applied to Services affect all consumers, preventing per-route customization
- **Extension limitations**: Current `BackendTLSPolicy` only supports Service targets, limiting external destination configuration

### Gateway API Integration Friction

Gateway API's backend reference system defaults to Services but:

> - **Limited extensibility**: Adding new backend types requires significant API changes

Reviewer 4: Can you summarize the significant changes?

Author: I actually don't like this entire section; removing

- **Inconsistent patterns**: Different implementations handle external destinations differently
- **Missing primitives**: No standard way to represent external FQDNs, IPs, or connection metadata

## Goals

> - **Introduce Backend resource** as a namespace-scoped consumer resource for representing destinations and connection metadata

Reviewer 4: This is a great goal, but the only motivation you've provided has to do with external destinations.  If what you're really doing here is abstracting the destination resource, there has to be some discussion on the motivation, limitations, and requirements for abstracting the destination resource.

Author: I think this is addressed now.

> - **Support multiple destination types**: Kubernetes Services and external FQDNs

Reviewer 4: This goal is better expressed as "Add one more destination type".   We already support Services.

Author: Reworked this section

- **Support producer-side policies**: Backend resource is explicitly consumer-focused
- **Standardize service mesh patterns**: Focus on Gateway API egress and decoration use cases
- **Provide cluster-scoped backends**: Backend resource is namespace-scoped for security boundaries
- **Address service discovery concerns**: Service discovery and DNS remain with existing Kubernetes Service system

## User Stories

### As an Application Developer

> "I want to configure my application to call external APIs (like OpenAI) with specific TLS settings and authentication without creating synthetic Services that expose security risks or affect other applications."

### As a Platform Engineer

> "I want to enforce that all external API calls go through specific gateways with proper logging and policy enforcement, without having to manage complex Service configurations for every external dependency."

### As a Security Administrator

> "I want to avoid ExternalName Services due to DNS rebinding vulnerabilities while still allowing applications to declare their external dependencies in a structured, auditable way."

## Proposal

> The Backend resource addresses a fundamental gap in Gateway API's backend reference system: first-class support for external destinations. Currently, external APIs, databases, and services must be represented as synthetic Kubernetes Services with `type: ExternalName`, which creates security vulnerabilities, policy application challenges, and resource management overhead. Furthermore, `Backend`s of type `EndpointSelector` can be used to decorate existing `Service`s (or other resources that fulfill the backend role) with TLS and protocol configuration.

Reviewer 4: Please remove EndpointSelector from this proposal until it is formally accepted.  To keep track, it could be added in a "Future Work" section.

```suggestion
The Backend resource addresses a fundamental gap in Gateway API's backend reference system: first-class support for external destinations. Currently, external APIs, databases, and services must be represented as synthetic Kubernetes Services with `type: ExternalName`, which creates security vulnerabilities, policy application challenges, and resource management overhead. 
```

Author: I would prefer not to. There has been enough consensus in Slack, multiple Gateway API meetings, as well as other subprojects/working groups (e.g. Agentic Networking, AI Gateway) to the point where it makes sense to progress these GEPs together

The Backend resource is explicitly designed as a **consumer resource** - it describes how a gateway should connect to a destination from the client perspective, regardless of whether that destination is internal or external to the cluster.

## API Specification

### Backend Resource Schema

```go
type Backend struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`
  Spec   BackendSpec   `json:"spec"`
  Status BackendControllerStatus `json:"status,omitempty"`
}

// +kubebuilder:validation:Enum=ExternalHostname;EndpointSelector
type BackendType string

const (
  BackendTypeExternalHostname             BackendType = "ExternalHostname"
  BackendTypeEndpointSelector             BackendType = "EndpointSelector"
)

> // +kubebuilder:validation:ExactlyOneOf=ExternalHostname,EndpointSelector

Reviewer 4: ```suggestion
```

Author: Why? Is it in the wrong place?

Reviewer 4: Isn't it redundant with 
// +kubebuilder:validation:Enum=ExternalHostname;EndpointSelector ?

Author: Ah I see; TIL. Thanks!

type BackendSpec struct {
  // Type defines the destination type

>   Type BackendType `json:"type"`

Reviewer 4: ```suggestion
  // +required
  Type BackendType `json:"type"`
```

>   // Ports defines the destination port for this backend.

Reviewer 4: Can you clarify what a destination port for a backend is and why it is required?

Author: Made an attempt; let me know what yuou think!

  // +required
  Port PortNumber `json:"ports,omitempty"`

  // ExternalHostname specifies the configuration for an ExternalHostname backend. Only used if type is ExternalHostname.
  // Support: Extended
  // +optional

>   // +optional

Author: ```suggestion
```

  ExternalHostname *ExternalHostnameBackend `json:"hostname,omitempty"`

  // EndpointSelector specifies the configuration for an EndpointSelector backend. Only used if type is EndpointSelector.
  // As defined in GEP-4731, creation of a `Backend` of type `EndpointSelector` should result in `Backend` controllers
  // creating the requisite `EndpointSelector` resource and setting ownerReferences appropriately.
  // TODO: Add link when GEP-4731 merges.
  // +optional
  EndpointSelector *EndpointSelectorBackend `json:"endpointSelector,omitempty"`

  // Protocol defines the higher-level protocol for backend communication.

>   // The underlying transport protocol for the proxied traffic will already have been

Reviewer 4: Missing an ending on this comment.

  // Support: Extended
  // +optional

>   // +kubebuilder:validation:MaxLength=256

Reviewer 4: Starting with a slice for potential multiple values would be helpful.  MaxLength 256 for one protocol seems excessive.

Author: Something went wrong in a rebase; I don't think we ever had a list of protocols. Thanks for the catch!

  Protocol BackendProtocol `json:"protocol"`

  // TLS defines the TLS configuration that a client should use when talking to the backend.
  // N.B: ExternalHostname backends SHOULD have TLS configured; the lack of TLS for external hostnames
  // should be considered insecure and a security risk.
  // +optional
  TLS *BackendTLS `json:"tls,omitempty"`
}

> // BackendTLSMode defines the TLS mode for backend connections.

Reviewer 4: We use the phrase "tls mode" in other places for something entirely different.  How about TLSDirection - unidirectional and bidirectional?

Author: I don't love the name, but I do agree that we need another look at the TLS setup here. I'm actually starting to come around to @snorwin's original suggestion [here](https://github.com/kubernetes-sigs/gateway-api/pull/4488#discussion_r3098214775) that we may be able to elide `mode` altogether. Ultiimately, the only change in consumer behavior is whether or not to attach a client cert

Author: Actually, with a None mode, we need some field after all. The structs and location are different so I think the naming is ok since the other usage is more for the frontend TLS case IIRC

// +kubebuilder:validation:Enum=ServerOnly;ClientAndServer
type BackendTLSMode string

const (
  // Enable TLS with simple server certificate verification.
  BackendTLSModeServerOnly BackendTLSMode = "ServerOnly"
  // Enable mutual TLS.
  BackendTLSModeClientAndServer BackendTLSMode = "ClientAndServer"
)

> type EndpointSelectorBackend struct {

Author: TODO: Add actual selector clause in there

  // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
  // conditions represent the current state of the Backend resource.
  // Each condition has a unique type and reflects the status of a specific aspect of the resource.
  //
  // Standard condition types include:
  // - "Available": the resource is fully functional
  // - "Progressing": the resource is being created or updated
  // - "Degraded": the resource failed to reach or maintain its desired state
  //
  // The status of each condition is one of True, False, or Unknown.
  // +listType=map
  // +listMapKey=type
  // +optional
  Conditions []metav1.Condition `json:"conditions,omitempty"`
}
```

### ExternalHostname Backend Configuration

```go
type ExternalHostnameBackend struct {
  // Hostname specifies the destination address used to reach this hostname.
  // IP addresses are not allowed in this field (enforced by validation on the type).
  // TODO: Should we add additional validation to prevent *.cluster.local hostnames?
  Hostname PreciseHostname `json:"address"`
}
```

### Protocol and Extension Support

```go

> // +kubebuilder:validation:Enum=MCP

Reviewer 4: Maybe add A2A, ANP, ACP?

Author: I'm wary about adding too protocols at the start. I'd like to get those and LLM in eventuallly, but want to start small to prove out the concept first

Reviewer 5: Agree with Keith, let's start simple ... I'm also worried about the mix of layers, please be sure we don't start to create conflicting configs, as per example we add JSON-RPC 2.0 and start going up and down in the stack

Author: @aojea yeah that's a definitely a concern. We can start with a small set of use-cases, prove them out while this is in experimental and then add others

   - Much simpler for implementations to integrate due to TLS settings being colocated with destination

2. **BackendTLSPolicy Limitations**
   - Current `BackendTLSPolicy` is designed around Service-based backends only
   - `BackendTLSPolicy` currently does not support per-consumer overrides
   - It is unclear whether `BackendTLSPolicy` is a producer or consumer oriented resource
     - [GEP 3875](https://github.com/kubernetes-sigs/gateway-api/pull/3876) proposed, among other things, adding consumer overrides to `BackendTLSPolicy`; however, that proposal introduced several new fields to the resource, including a `from` selector on `targetRef` that would have added significant complexity to both the API and implementations. Furthermore, the GEP has a [limitation](https://github.com/kubernetes-sigs/gateway-api/pull/3876/changes#diff-67a0076fb272af6273ce353d4687732735e03ddec8ae2bbc35b0a905281f9057R88) that it would not be possible to enforce consumer-side policies originating from the same namespace as the producer. This proposal was ultimately abandoned.

3. **Per-Backend Client Certificates**

>    - External destinations may require different client certificates

Reviewer 4: Different than what?

Author: Reworded

>    - Current Gateway API patterns only support one client certificate per Gateway

Reviewer 4: Only in the case where the Gateway is acting as a client.  Passthrough TLS mode allows more than one, if I'm not mistaken.

   - Backend-specific client certificates are essential for many external integrations

#### Arguments for Policy-Based Configuration

1. **API Consistency**
   *Note: the following points, while true, are only true because `Service` is effectively immutable.*
   - Gateway API uses policy attachment for most configuration beyond basic routing
   - Inline configuration creates another place to define TLS settings
   - Multiple places for TLS configuration could lead to misalignment of features

2. **Reusability and Standardization**
   - Policies can be shared across multiple Backends
   - Consistent TLS configuration patterns across resource types (e.g. `Service`, `Backend`, `InferencePool`, etc.)

#### Proposed Approach

3. **Per-Backend Client Certificates**
   - External destinations may require different client certificates

>    - Current Gateway API patter

Reviewer 4: This sentence is missing an ending.

Author: Something went weird in a rebase; fixed

Based on community feedback and practical considerations, this proposal recommends:

1. **Inline TLS for Backend Resources**: Provide inline TLS configuration within the Backend resource for simplicity and external destination requirements

2. **Explicit BackendTLSPolicy Exclusion**: Backend resources are explicitly disallowed as targets for `BackendTLSPolicy` to avoid confusion and conflicts

3. **Type Definition Alignment**: Align the inline TLS types with `BackendTLSPolicy` types as closely as possible for consistency

### Implementation Example

```yaml
# Gateway-level TLS remains authoritative for incoming connections
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
spec:
  listeners:
  - name: https
    protocol: HTTPS
    tls:
      certificateRefs:
      - name: gateway-cert

>   # Deprecate gateway.spec.tls.backend in favor of Backend resource configuration

Reviewer 4: gateway.spec.tls.backend is Core.  Doesn't that mean we can't deprecate it?

Author: Old draft; good catch

---
# Backend resource with inline TLS for external destination
apiVersion: gateway.networking.k8s.io/v1alpha1
kind: Backend
metadata:
  name: openai-api
  namespace: ai-apps
spec:

>   destination:

Reviewer 4: The `destination` field doesn't appear in the API discussion.

Author: Fixed; it was from an old revision

    type: ExternalHostname
    externalHostname:
      hostname: api.openai.com
      tls:
        serverName: api.openai.com
        clientCertificate:
          secretRef:
            name: openai-client-cert
            key: tls.crt

>     ports:

Reviewer 4: ```suggestion
    externalHostname:
      hostname: api.openai.com
    tls:
      mode: BackendTLSModeClientAndServer
      clientCertificateRef:
        secretRef:
          name: openai-client-cert
          key: tls.crt
   port:
```

    - name: https
      port: 443

>       protocol: HTTPS

Reviewer 4: ```suggestion
      name: https
      port: 443
      protocol: HTTPS
```

   - Restrictive validation would break legitimate external integrations
   - Security focus should be on network-level controls, not resource-level validation
   - No initial support for wildcard hostnames to limit attack surface and the need for Dynamic Forward Proxy support from implementations
     - Future proposals to add this functionality should include a comprehensive DNS trust specification and threat model.
   - Implementations may also be able to implement data-plane/proxy-level protections for common attack vectors
   - NOTE: This may be a decision we revisit in the future based on user feedback

#### Example of a Recommended Network Policy

```yaml
# Network policy enforcement
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: egress-through-gateway
  namespace: application-team
spec:
  podSelector: {}
  policyTypes:
  - Egress
  egress:
  - to:
    - namespaceSelector:
        matchLabels:
          name: gateway-system
    ports:
    - protocol: TCP
      port: 8080  # Gateway proxy port

>   # Block direct external egress, force through gateway

Reviewer 4: Is this comment a TODO, a recommendation, or an explanation of what's happening here?

Author: Description; I'll clarify (and correct the NetworkPolicies because they were wrong)

The Backend resource is explicitly designed as a **consumer resource** - it describes how a gateway should connect to a destination from the client perspective, regardless of whether that destination is internal or external to the cluster.

## API Specification

### Backend Resource Schema

```go
type Backend struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`
  Spec   BackendSpec   `json:"spec"`
  Status BackendControllerStatus `json:"status,omitempty"`
}

type BackendSpec struct {
  // Destination defines where traffic should be sent
  Destination BackendDestination `json:"destination"`

  // Filters defines filters that should be executed when
  // sending traffic to this backend. Filters should not
  // be duplicated on a backendRef (targeting this `Backend`)
  // and on the `Backend` itself.
  // +optional
  // TODO: Specify filter type definition. Should,
  // at minimum, include ExtensionRef pattern.
  Filters []BackendFilters `json:"filters,omitempty"`
}

type BackendType string

const (
  BackendTypeFQDN    BackendType = "FQDN"

>   // TODO: Add WorkloadSelector destination type

Author: Added a reference.

)

type BackendDestination struct {
  // Type defines the destination type
  Type BackendType `json:"type"`

  // Ports defines the destination ports and protocols
  // kubebuilder:validation:MinItems=1
  // kubebuilder:validation:MaxItems=16
  Ports []BackendPort `json:"ports,omitempty"`

  // FQDN specifies an external fully qualified domain name
  // +optional
  FQDN *FQDNBackend `json:"fqdn,omitempty"`
}

// BackendTLSMode defines the TLS mode for backend connections.
// +kubebuilder:validation:Enum=Simple;Mutual;None
type BackendTLSMode string

const (
  // Do not modify or configure TLS. If your platform (or service mesh)
  // transparently handles TLS, use this mode.
  BackendTLSModeNone BackendTLSMode = "None"
  // Enable TLS with simple server certificate verification.
  BackendTLSModeSimple BackendTLSMode = "Simple"
  // Enable mutual TLS.

>   BackendTLSModeMutual BackendTLSMode = "Mutual"

Reviewer 6: To differentiate `Mutual` from `Simple` TLS, I think checking whether `clientCertificateRef` was specified is sufficient. Explicitly requiring the user to configure a `mode` adds little value over inferring it implicitly.

>   BackendTLSModeMutual BackendTLSMode = "Mutual"

Author: Hmm I don't love the idea of inferring the user's TLS intent based purely on the presence/absence of clientCertRefs. We can't  force it to not be nil so how does one interpret SimpleTLSMode with an  client certificate ref explicitly set to nil (vs omitted)? I think being explicit with user intent is better here

>   BackendTLSModeMutual BackendTLSMode = "Mutual"

Reviewer 2: I agree that a `mode` selector is probably a necessary evil, but I would argue that we should name the modes more literally - perhaps `ServerOnly`, and `ClientAndServer`. Again, "mutual TLS" _can_ mean "client presents a certificate", or it _can_ mean "client and server both hold centrally-managed certificates that are frequently rotated by some other mechanism". In this case, we mean the less-specific former, rather than the implicit behavior of the latter, so I don't like the possibility of conflating the concerns.

For better or for worse, when end users see "Mutual TLS", or "mTLS", they think "service mesh" and container identity as much as anything else. So we need to ensure that this usage is clearly different to that.

>   BackendTLSModeMutual BackendTLSMode = "Mutual"

Author: I can get behind that

}

type BackendPort struct {
  // Number defines the port number of the backend.
  // +required
  // +kubebuilder:validation:Minimum=1
  // +kubebuilder:validation:Maximum=65535
  Number uint32 `json:"number"`
  // Protocol defines the protocol of the backend.
  // +required
  // +kubebuilder:validation:MaxLength=256
  Protocol BackendProtocol `json:"protocol"`
  // TLS defines the TLS configuration that a client should use when talking to the backend.
  // TODO: To prevent duplication on the part of the user, maybe this should be declared once at the
  // top level with per-port overrides?
  // +optional
  TLS *BackendTLS `json:"tls,omitempty"`
  // +optional
  ProtocolOptions *BackendProtocolOptions `json:"protocolOptions,omitempty"`
}
```

### FQDN Backend Configuration

```go
type FQDNBackend struct {
  // Hostname specifies the destination FQDN
  Hostname string `json:"hostname"`
}
```

### Protocol and Extension Support

```go

// BackendProtocol defines the protocol for backend communication.
// +kubebuilder:validation:Enum=HTTP;HTTP2;TCP;MCP
type BackendProtocol string

const (
  BackendProtocolHTTP  BackendProtocol = "HTTP"

>   BackendProtocolHTTP2 BackendProtocol = "HTTP2"

Reviewer 3: As a user, I probably have no clue what version `openai.com` (or any hostname) supports. That is what ALPN is for, so I don't have to 🙂.

So I think at minimum clarifying in the spec what `HTTP` means is important, and possibly improvements to allow a negotiated HTTP version

>   BackendProtocolHTTP2 BackendProtocol = "HTTP2"

Author: Do we even need protocol? If this is always going to be attached to a route (or future `Frontend`), the protocol is already set

>   BackendProtocolHTTP2 BackendProtocol = "HTTP2"

Author: Removed HTTP and TCP in favor of relegating protocol to solely describing any higher-level protocols the backend might be expecting (or potentially codifying rewrites or semantics like MCP path or OpenAI/Anthropic conventions). Added documentation that explains that the initial L4/L7 processing of the gateway will be decided by the route

>   BackendProtocolHTTP2 BackendProtocol = "HTTP2"

Author: I ended up adding almost all of these back for the validation and upgrade/downgrade use-case

        serverName: api.openai.com
        clientCertificate:
          secretRef:
            name: openai-client-cert
            key: tls.crt
    ports:
    - name: https
      port: 443
      protocol: HTTPS

---
# HTTPRoute referencing Backend
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
spec:
  rules:
  - backendRefs:
    - name: openai-api
      kind: Backend
      group: gateway.networking.k8s.io
```

## Security Model and RBAC Considerations

### FQDN Security Analysis

Allowing namespace-scoped Backend resources to reference external FQDNs raises legitimate security concerns that must be carefully considered:

#### Identified Security Risks

1. **DNS Spoofing Attacks**
   - Malicious DNS responses could redirect traffic to attacker-controlled servers
   - Particularly concerning for internal proxy endpoints or localhost addresses (i.e. the [Confused Deputy Problem](https://en.wikipedia.org/wiki/Confused_deputy_problem))
   - Risk: `api.external.com` resolves to `127.0.0.1`, `169.254.169.254` or other privileged, trusted addresses

2. **Cross-Namespace Service Access**
   - FQDNs could target internal cluster services via `svc.namespace.svc.cluster.local`

>    - Potential bypass of namespace isolation and RBAC controls

Reviewer 7: Not sure I follow this? RBAC does not control traffic Authz to other servers. Do you mean authz policies and not k8s rbac?

This conflation creates several significant issues:

### External Destination Limitations

Currently, representing external destinations in Gateway API requires synthetic `Service` objects with `type: ExternalName`, which:

- **Security vulnerabilities**: ExternalName Services are subject to DNS rebinding attacks ([CVE-2021-25740](https://github.com/kubernetes/kubernetes/issues/103675))
- **Policy limitations**: Cannot apply backend-specific policies (TLS, authentication, rate limiting) without affecting all consumers
- **Synthetic resource overhead**: Creates artificial Kubernetes resources for external dependencies
- **Configuration complexity**: Mixing internal and external destinations in the same resource type

### Policy Application Complexity

The overloaded Service resource makes policy attachment ambiguous:

- **Producer vs Consumer confusion**: Services represent both "how to call me" (producer) and "where to route" (consumer) concerns
- **Policy scope unclear**: TLS policies applied to Services affect all consumers, preventing per-route customization

> - **Extension limitations**: Current `BackendTLSPolicy` only supports Service targets, limiting external destination configuration

Reviewer 6: That's not entirely correct, BackendTLSPolicy supports other targets as well. However, since we aren't aware of targets that would be broadly relevant across many implementations, supporting targets other than services referenced from an HTTPRoute is considered implementation-specific.

Even if we could use BackendTLSPolicy for the backend, I would strongly suggest inlining the TLS configuration rather than relying on BackendTLSPolicy, in favor of reducing configuration complexity.

https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/backendtlspolicy_types.go#L98

Author: Fair point. I'll reword or remove

Reviewer 7: > in favor of reducing configuration complexity.

@snorwin how does this *reduces configuration complexity* ? If you have multiple ways to do TLS within the ecosystem doesn't it adds for configuration complexity?

Reviewer 6: Based on the feedback we've received, users perceive it as more complex when a single intent (e.g., my backend is https://my-backend.io) has to be expressed across multiple CRs.
This was actually one of the main criticisms from people migrating from ingress-nginx to Gateway API, having to deal with multiple CRs to express the same configuration. 

IMO, policy attachment makes sense when we don't control the resource a policy would be attached to, or when a policy can be attached to multiple resources.

Reviewer 2: Thanks @snorwin, I agree. If we're adding new things that effectively wrap Service, we should definitely bring as much functionality as we need out of Policy and into that thing instead. BackendTLSPolicy won't go away now, because it's Stable, but if we have a better option in Backend, then I think it's an acceptable tradeoff to put the TLS config there as well.

Reviewer 4: IMO we are rushing to do too many things at once in this proposal without a clear motivation. Let's separate the two concerns (egress and decoration use cases) and tackle them individually.

Author: I wouldn't say that we're rushing; this GEP is almost 3 months old at this point....when designing egress, the question naturally arose: "don't we want to do all of these things for Services too?". In fact, the agentic networking group was going to create their own XBackend resource because they needed to be able to decorate `Service`. Adding EndpointSelector into scope for this GEP is meant, among other things, to prevent 3-4 different Backend resources/workarounds popping up in various subprojects

// +kubebuilder:validation:Enum=Simple;Mutual;None
type BackendTLSMode string

const (
  // Do not modify or configure TLS. If your platform (or service mesh)
  // transparently handles TLS, use this mode.
  BackendTLSModeNone BackendTLSMode = "None"
  // Enable TLS with simple server certificate verification.
  BackendTLSModeSimple BackendTLSMode = "Simple"
  // Enable mutual TLS.
  BackendTLSModeMutual BackendTLSMode = "Mutual"
)

type EndpointSelectorBackend struct {
  // SelectorRef specifies the reference to the EndpointSelector resource that manages the EndpointSlices for this backend.
  // +required
  SelectorRef *LocalObjectReference `json:"selectorRef"`
}

type BackendTLS struct {
  // Mode defines the TLS mode for the backend.
  // +required
  Mode BackendTLSMode `json:"mode"`

  // ClientCertificateRef defines the reference to the client certificate for mutual
  // TLS. Only used if mode is MUTUAL.
  // +optional
  ClientCertificateRef *SecretObjectReference `json:"clientCertificateRef,omitempty"`

  // Re-use BackendTLS policy validation fields. This is currently missing InsecureSKipVerify
  // but that will be added in GEP-4152.

>   Validation BackendTLSPolicyValidation `json:"validation,omitempty"`

Reviewer 6: @keithmattix can you please check https://github.com/kubernetes-sigs/gateway-api/pull/4779 and verify if this work for your use cases?

Author: Yes it does; thank you!

  // TODO: To prevent duplication on the part of the user, maybe this should be declared once at the
  // top level with per-port overrides?
  // +optional
  TLS *BackendTLS `json:"tls,omitempty"`
  // +optional
  ProtocolOptions *BackendProtocolOptions `json:"protocolOptions,omitempty"`
}
```

### Hostname Backend Configuration

```go
type HostnameBackend struct {
  // Address specifies the destination address used to reach this hostname.
  Address string `json:"address"`
}
```

### Protocol and Extension Support

```go

// BackendProtocol defines the protocol for backend communication.
// +kubebuilder:validation:Enum=HTTP;HTTP2;TCP;MCP
type BackendProtocol string

const (
  BackendProtocolHTTP  BackendProtocol = "HTTP"
  BackendProtocolHTTP2 BackendProtocol = "HTTP2"
  BackendProtocolTCP   BackendProtocol = "TCP"
  BackendProtocolMCP   BackendProtocol = "MCP"
)

// +kubebuilder:validation:ExactlyOneOf=mcp
type BackendProtocolOptions struct {
  // +optional
  MCP *MCPProtocolOptions `json:"mcp,omitempty"`
}

type MCPProtocolOptions struct {
  // URL path for MCP traffic. Default is /mcp.
  // +optional
  // +kubebuilder:default:=/mcp

>   Path string `json:"path,omitempty"`

Reviewer 7: Whats the semantics of this field? No matter how request gets to this backend (i.e with path "/keith" - would it path-rewrite it to this path?)?

Author: Removing path in favor of a standalone GEP

}
```

## TLS Policy Consolidation Analysis

One of the most significant design decisions for the Backend resource concerns TLS configuration: should it be inline within the Backend resource or provided through separate policy resources like `BackendTLSPolicy`?

### Tradeoffs: Inline TLS vs. Policy-Based TLS

#### Arguments for Inline TLS Configuration

1. **Simplified UX for External Destinations**
   - External FQDNs often require TLS configuration that is specific to that destination
   - Better discoverability for users who want to understand what TLS settings apply for backends within a route
   - Much simpler for implementations to integrate due to TLS settings being colocated with destination

2. **BackendTLSPolicy Limitations**
   - Current `BackendTLSPolicy` is designed around Service-based backends only

>    - `BackendTLSPolicy` currently does not support per-consumer overrides

Reviewer 7: At some point in the past, I brought up John's PR to enhance BackendTLSPolicy to support per-consumer overrides. I can try digging it up again, but why it is not listed here? I think you added it at some point

Author: I just thought we discussed it enough synchronously. IIRC, the idea was that inline TLS is simpler (it's what we would do with Service if we could) and the change in semantics based on kind is actually fairly complex for users

Reviewer 7: see below comment. I believe we said this GEP will mentions Johns proposal and the notes we had on that, and will explain why its inferior. Same thing for BTLSPolicy targets backend Backend

Author: Sure I can add it

Author: Added a bullet point

>    - It is unclear whether `BackendTLSPolicy` is a producer or consumer oriented resource

Reviewer 7: We have talked in multiple WG AI GW meetings -- there is another alternative which is `BTLSPolicy --> Service == producer policy` and `BTLSPolicy --> Backend == consumer policy`

Can we compare these options?

Author: Yeah we've discussed this a couple of times, but I think the consensus has formed around inline policy at this point.

Reviewer 7: IIRC i think the result of the discussions I was part of is that this GEP will have mention this option and explain why its a less compelling alternative. 

The way this GEP is written is that if this option does not exist. 

(tightly related to this comment https://github.com/kubernetes-sigs/gateway-api/pull/4488#discussion_r3132679720 but not exactly the same)

Author: Added a bullet point with a link to the GEP and why we're not reviving it

- Backend resources provide audit trail for external dependencies

## EndpointSelector Type

## Extension Framework

The Backend resource provides three levels for applying extensions and policies:

### 1. Route-Level Extensions (HTTPRoute Filters)

Applied to individual requests as they are routed to a Backend.

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
spec:
  rules:
  - matches:
    - path: { value: "/api/models" }
    filters:
    - type: ExtensionRef
      extensionRef:
        name: rate-limiter
        kind: RateLimitPolicy
    backendRefs:
    - name: openai-api
      kind: Backend
```

**Use cases**: Request modification, rate limiting, authentication injection

### 2. Backend-Level Extensions

Applied at the Backend resource, affecting all requests to that destination.

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha1
kind: Backend
spec:
  destination:
    type: Hostname
    hostname:
      address: api.openai.com
  filters:
  - type: ExtensionRef
    extensionRef:
      name: connection-pool
      kind: ConnectionPoolPolicy
  - type: ExtensionRef
    extensionRef:
      name: circuit-breaker
      kind: CircuitBreakerPolicy

> ```

Reviewer 7: are these even implementable in practice for common dataplanes today? e.g for envoy not everything is supported as upstream filters

Author: Discussed in WG AI Gateway; we're going to remove filters and probably make it its own GEP

* Issue: [#4488](https://github.com/kubernetes-sigs/gateway-api/pull/4488)
  * Incubated by the [AI Gateway Working Group](https://github.com/kubernetes-sigs/wg-ai-gateway/pull/20)
* Status: Experimental

## TLDR

This GEP proposes a new `Backend` resource that fills the [backend role](https://gateway-api.sigs.k8s.io/geps/gep-2907/) — a **general-purpose decorator for Service** (and other backend types) within Gateway API. The Kubernetes `Service` resource is mature and stable, but it is effectively frozen and SIG-Network leadership has an extreme aversion to expanding its scope. Previous approaches to extend Service behavior (like `BackendTLSPolicy`) have significant limitations around discoverability, implementation complexity, and the conflation of producer and consumer concerns.

> The `Backend` resource provides a namespace-scoped, consumer-focused resource that can:

Reviewer: Is the goal to migrate all service-attached policies to Backend? I'm curious what happens with producer-oriented configuration, like `retryConstraint` (https://gateway-api.sigs.k8s.io/geps/gep-3388/). Is there a future for producer policy on Backend?

My concern is that one of the main push for Backend is to solve the policy problem, but at the same time, Backend is consumer-only. 

CC: @youngnick if you have any thoughts on this, I want to make sure I understand the direction here.

Reviewer 2: Yes, I think that part of this change is that Backend is _not_ consumer focussed any more. It can be used for consumer uses - in the ExternalName case, mainly, but other things we will put there may be more producer focussed, like Session Persistence.

Maybe we can make it clearer as part of the Backend design by having `consumer` and `producer` stanzas or something, to avoid the lack of clarity we have in Service.

Reviewer 2: To be more clear and specific:

I think that Backend is useful in both the consumer and producer contexts. In the consumer context, it provides settings for the Gateway to consume directly. In the producer context, it provides _hints_ from the producer that allow a consumer to determine, _a priori_ if it's worth trying to connect at all, or what shape a _successful_ connection would look like. The producer settings _often_ overlap with the consumer settings, but not always.

Session persistence is a great example here, because session persistence _requires_ that the application be session aware in some fashion, and the configuration for the consumer is about making sure that the consumer's expectations match the producers. If that's not the case, then the session persistence just won't work. So it would be _really_ useful to be able to have a place where producers/servers can say "hey, I expect session persistence to look like this", and _maybe_ also a place to for consumers/clients can say "oh, I can't do that, this is guaranteed to fail" or something.

Author: I actually want to be very careful here because one of the things that makes Backend's scope feasible is that we can limit it to consumer use-cases. Even ExternalHostname functionality is still consumer, dictating how the external service is accessed.

Reviewer 2: There is a _lot_ of utility in having something like Backend that can handle producer cases, like Session Persistence and the server side of TLS.

That _could_ be Backend, if we are clear about where the respsonsibility lies. ExternalName support is clearly a consumer thing, but if we do want to add a Service wrapper that handles the consumer side, what we call it other than Backend? Is Backend really ConsumerBackend, and we need another thing to be ProducerBackend? 

I understand the use cases most folks are thinking about _right now_ are consumer ones, but as I have been saying, the producer cases are also important. Session Persistence is a great one, but one of the reasons BackendTLSPolicy has been a confusing and argumentative process is the constant confusion over whether it's a consumer or producer config.

If Backend is always and only going to be consumer-focussed, then we can't use such a generic name, because I won't be the only one who assumes it's for more than just consumer uses.

Author: If we go back to the role definitions, isn't the backend/bag of endpoints role not purely consumer-oriented? You're choosing which endpoints to send to. Maybe I'm just missing something

Reviewer 2: It's more about "who owns the HTTPRoute, and do they own the service they are sending traffic to?"

Most of the initial Ingress work has been done on the assumption that Ana owns _both_ the HTTPRoute _and_ the backends it's sending to, so only the bag of endpoints is really relevant in that case.

But in the case of Backend with ExternalName, it's _explicitly_ for when the owner of the HTTPRoute does not control the backend. Maybe it's a bag of endpoints inside the cluster, maybe it's outside the cluster. But the HTTPRoute owner is _not_ the same person as the endpoint owner.

However, a construct like Backend is relevant for the person who owns the endpoints (ie the producer), because it allows them to specify hints/requirements for connection that will apply regardless of where that Backend is used. Session persistence and the _server_ parts of BackendTLSPolicy are relevant here. The Gateway implementation can _often_ look at the producer requirements and use them, but not always. (Session persistence "I expect a session key in this header" is a good example of the former, and "I accept only TLS v1.1" is an example of the latter).

That's what I mean by personas, there's an implicit assumption in the Ana persona that she (or a team represented by her) controls _both_ the HTTPRoute and the endpoints it sends traffic to, most of the time. Ana owns an Application, that has Pods that expose Endpoints, that are grouped with a Service, and exposed with a HTTPRoute that connects to a Gateway.

I think that an agentic or AI user (let's call her Greta, since she's either an aGent or is using an aGent, so the name should start with G) is different in that, while they _may_ operate a server (an MCP server or model) inside the cluster, it's also imperative that the user be able to have access to _applications they don't own_ managed. So the HTTPRoute that this agentic persona owns is different in that it's pointing at an application that Greta doesn't own, either in the cluster (which would use an endpointselector or Service reference or something), or outside the cluster (which would use an ExternalName/hostname). It's definitely useful to be able to do this sort of consumer control, but it's _way_ closer to GAMMA consumer policies than it is to Gateway for Ingress.

I think that slight nuance in the ownership and conceptual model makes it difficult to talk about this without talking past each other.

I'm way more prone to assuming the Ana-model than I am the Greta-model, because I've been focussed on it for 7 years, and am still mainly focussed on it.

@keithmattix and folks who've been focussed on service mesh and AI use cases for some time now tend to assume the Greta-model of ownership, and this difference leads to starting from a different place, which leads to weird communications breakdowns.

So yeah, what I'm trying to say is:

* I think there are definitely clear use cases for Greta, who owns HTTPRoutes that connect to backends she doesn't control (generally)
* But, there are also places where having a single object for producer, Ana-centric use cases, where Ana owns the endpoints behind the Routes, and _may_ have her endpoints used in many HTTPRoutes.

Previously, we've tried to create things that handled some of this, and had a _lot_ of confusion over whether it's consumer or producer config. BackendTLSPolicy is the _prime_ example here. It started out being conceived of as producer-only config, but ended up being mainly consumer config in its use and design (after implementations that assumed it was consumer config added more features).

What I'm trying to do for Backend is to make this clearer upfront. Is this always and only for consumer things? Or could it be used for producer things as well? If the latter, how do we make the behavior clear? If the former, is a generic name like Backend really okay, since this is actually less generic than that name implies?

Author: @youngnick i think that's really well said and captures the nuance perfectly. Indeed, in mesh and egress use-cases, the ownership of the route relative to the backend is different than ingress (oftentimes there isn't even a route!), so that's definitely been shaping my perspective throughout this design process; I'll take a much closer look at this

Author: The outcome of this thread and the discussion that followed can be found in the beginning TL;DR section of the GEP

* Issue: [#4488](https://github.com/kubernetes-sigs/gateway-api/pull/4488)
  * Incubated by the [AI Gateway Working Group](https://github.com/kubernetes-sigs/wg-ai-gateway/pull/20)
* Status: Experimental

## TLDR

> This GEP proposes a new `Backend` resource that fills the [backend role](https://gateway-api.sigs.k8s.io/geps/gep-2907/) — a **general-purpose decorator for Service** (and other backend types) within Gateway API. The Kubernetes `Service` resource is mature and stable, but it is effectively frozen and SIG-Network leadership has an extreme aversion to expanding its scope. Previous approaches to extend Service behavior (like `BackendTLSPolicy`) have significant limitations around discoverability, implementation complexity, and the conflation of producer and consumer concerns.

Reviewer 8: very nit: I would avoid using the words "extreme version" as they are "a lot extreme" but claim that extending services has implications on Services API that we don't want to be responsible, including adding conformance upstream for it.

Or leave it as is, but this wording sounded very strong to me.

Also we can say that BackendTLSPolicy was the right solution for the moment, but we figured out that policy attachment may not be the best approach for all the features a backend needs.

Author: Reworded, thanks!

The `Backend` resource provides a namespace-scoped, consumer-focused resource that can:

1. **Decorate existing Services** via `EndpointSelector`, adding Gateway-specific configuration (TLS, protocol, etc.) without modifying the Service itself.
2. **Represent external destinations** via `ExternalHostname`, replacing the need for insecure synthetic `ExternalName` Services.
3. **Serve as a foundation for future Gateway-level backend configuration** such as retries, session persistence, load balancing algorithms, and other features that are tightly bound to the destination rather than the route.

While egress and AI use cases provided the initial urgent motivation, the `Backend` resource is designed to be useful for **all backend types**. At its core, a `Backend` of type `EndpointSelector` does what `Service` already does — but in a Gateway-native way that allows configuration to grow over time. Egress support via `ExternalHostname` is the first Extended feature built on this foundation.

### Clarifying the Semantics of "Backend"

> It is critically important that we emphasize that the `Backend` resource describes what a specific gateway client connection **MUST** do on the wire when connecting to a destination.

Reviewer 5: naive question, client here is the Gateway itself or the client as a pod , user? I imagine the former but I just want to be sure if we are entering the realm of pod traffic here

Author: The gateway generally in the ingress use-case, but would also apply to a mesh proxy/dataplane. If you're thinking about whether the CNI or something like kube-proxy counts as a dataplane, I would say the answer is "hopefully in the future" but we'd need to be clear that all of the L7 fields (plus TLS) aren't implemented

In the common ingress persona, Ana often owns both the `HTTPRoute` and the Service behind it. In this scenario, Ana is being delegated the ability to control how the Gateway consumes her service. In many egress, mesh, and AI-oriented deployments, that ownership model changes: Ana does NOT own the destination behind the route, but she still needs a way to express "how should the gateway connect to this destination" without needing to coordinate with the producer or cluster admin. In other words, the `Backend` resource is a consumer-side resource that describes connection requirements for a particular client path, regardless of who owns the destination.

This distinction is foundational to this GEP:

- `Backend` captures consumer-side connection requirements for a particular client path (for example, SNI, TLS validation, client cert presentation, or higher-level protocol expectations).
- Producer/server guidance is still valuable, out of scope for this resource: producer hints describe what servers generally **SHOULD** accept, while this resource describes what this client/gateway connection **MUST** attempt.

> - In practice, server hints were always actuated as client configuration anyway. This proposal makes that behavior explicit and auditable.

Reviewer 5: My comment here is about the failure mode, apologize if is in other places but I couldn;t find so just feel free to redirect me directly.
IIUIC this means that users will express their intent but we know they are going to put wrong configurations, how is this going to be detected? it will fail only at runtime? 
Are gateway clients expected to "active validate" the consumer/producer and report back?

Author: At this point, yes it will fail only at runtime, and I think that's acceptable because it's no worse than how ~every connection on the internet works today. I've been talking with @kflynn, @youngnick, @mikemorris, @rikatz, and others on an enhancement to this area of resources that will allow producers to surface hints (e.g. min/max TLS version) that could help control planes detect misconfigurations statically

Reviewer 8: we can also split this concern into 2 different scenarios:
* A Gateway that does not support  a consumer intent on Backend configuration must surface it as a condition of the Backend. Saying on other words, let's say a user establishes "I want this Backend to use this cipherSuite to connect to the services" and the field CipherSuite is not supported, the implementation must add to the condition
* A user puts a bad configuration like "use this tlsCipher" but the tlsCipher is not supported by the backend, means the gateway will try to open a connection to the backend using that cipher and have an error. This is not different of today BackendTLSPolicy or other resources that establishes the behavior of runtime and runtime fails, which usually causes an error 500. This will not change (it is similar to selecting a service and the service is not responsive, there is no way on the control plane to test that configuration, just during user requests)

Author: For the first one, isn't our current approahc supportedFeatures? +1 on the second point

The Backend resource is explicitly designed as a **consumer resource** — it describes how a gateway should connect to a destination from the client perspective, regardless of whether that destination is internal or external to the cluster.

## API Specification

### Backend Resource Schema

```go
type Backend struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`
  Spec   BackendSpec   `json:"spec"`
  Status BackendControllerStatus `json:"status,omitempty"`
}

// +kubebuilder:validation:Enum=ExternalHostname;EndpointSelector
type BackendType string

const (
  BackendTypeExternalHostname             BackendType = "ExternalHostname"
  BackendTypeEndpointSelector             BackendType = "EndpointSelector"
)

> // +kubebuilder:validation:ExactlyOneOf=ExternalHostname,EndpointSelector

Reviewer: This enforces that either one is set, but should we enforce that the correct one is set? i.e. `spec.hostname` is set when `spec.type: externalHostname`, and `spec.endpointSelector` is set when `spec.type: EndpointSelector`.

Author: Great catch! Added

type BackendSpec struct {
  // Type defines the destination type
  // +required

>   Type BackendType `json:"type"`

Reviewer: Should this be a discriminated union? By convention, you'd need to change `hostname` to `externalHostname` to keep that pattern aligned and add some CEL to enforce it. HTTPRouteFilterType [Example](https://github.com/kubernetes-sigs/gateway-api/blob/5c120d1a4484966a8bf9ef5264646e9d3993ad04/apis/v1/httproute_types.go#L857)

Author: Yep, that's the idea. I think I fixed this.

  // Ports defines the port that the implementation should use when connecting to this backend.
  // +required

>   Port PortNumber `json:"ports,omitempty"`

Reviewer: *nit* 

```suggestion
  // Port defines the port that the implementation should use when connecting to this backend.
  // +required
  Port PortNumber `json:"port,omitempty"`
```

  // ExternalHostname specifies the configuration for an ExternalHostname backend. Only used if type is ExternalHostname.
  // Support: Extended
  // +optional

>   ExternalHostname *ExternalHostnameBackend `json:"hostname,omitempty"`

Reviewer: Is this field name intentionally `hostname`, while the docs and type are `ExternalHostname`?

Author: Yep good catch; missed it in the rename

  BackendTLSModeServerOnly BackendTLSMode = "ServerOnly"
  // Enable mutual TLS.
  BackendTLSModeClientAndServer BackendTLSMode = "ClientAndServer"
)

// +kubebuilder:validation:ExactlyOneOf=SelectorRef,Selector
type EndpointSelectorBackend struct {
  // SelectorRef specifies the reference to the EndpointSelector resource that manages the EndpointSlices for this backend.
  // If omitted, the controller creating this Backend is expected to create an EndpointSelector
  // resource on behalf of the user and set ownerReferences appropriately so that the lifecycle
  // of the EndpointSelector is tied to this Backend (as described in GEP-4731).
  // +optional
  SelectorRef *LocalObjectReference `json:"selectorRef"`

  // Selector defines the label selector used to identify the endpoints that this backend
  // should route traffic to. This field is only used if SelectorRef is not specified.
  // +optional
  Selector *metav1.LabelSelector `json:"selector,omitempty"`
}

type BackendTLS struct {
  // Mode defines the TLS mode for the backend.
  // +required
  Mode BackendTLSMode `json:"mode"`

  // ClientCertificateRef defines the reference to the client certificate for mutual

>   // TLS. Only used if mode is MUTUAL.

Reviewer: *nit*

```suggestion
  // TLS. Only used if mode is ClientAndServer.
```

The Backend resource is explicitly designed as a **consumer resource** — it describes how a gateway should connect to a destination from the client perspective, regardless of whether that destination is internal or external to the cluster.

## API Specification

### Backend Resource Schema

```go
type Backend struct {
  metav1.TypeMeta   `json:",inline"`
  metav1.ObjectMeta `json:"metadata,omitempty"`
  Spec   BackendSpec   `json:"spec"`

>   Status BackendControllerStatus `json:"status,omitempty"`

Reviewer 8: nit: why not just BackendStatus to reflect BackendSpec?

Author: This is to indicate that it's per controller status

Reviewer 8: commented directly on Slack, we probably must follow <Kind>Status here

And the status probably needs a []ParentRef or something similar if we want multiple controllers

Author: Nice catch; done

}

// +kubebuilder:validation:Enum=ExternalHostname;EndpointSelector
type BackendType string

const (
  BackendTypeExternalHostname             BackendType = "ExternalHostname"
  BackendTypeEndpointSelector             BackendType = "EndpointSelector"
)

// +kubebuilder:validation:XValidation:rule="self.type == 'ExternalHostname' ? has(self.externalHostname) : !has(self.externalHostname)",message="externalHostname must be set when type is ExternalHostname and must be unset otherwise"
// +kubebuilder:validation:XValidation:rule="self.type == 'EndpointSelector' ? has(self.endpointSelector) : !has(self.endpointSelector)",message="endpointSelector must be set when type is EndpointSelector and must be unset otherwise"
type BackendSpec struct {
  // Type defines the backend type
  //
  // +unionDiscriminator
  // Port defines the port that the implementation should use when connecting to this backend.
  // +required
  // Port defines the port that the implementation should use when connecting to this backend.

>   // +required

Reviewer 8: something weird here, the field doc is duplicated and the markers are on some strange order

Author: Yeah good catch; I don't know what happened here (maybe a rebase went bad?)

>   Port PortNumber `json:"port,omitempty"`

Reviewer 8: do we care about named ports in case backendRef is a service? or even better, do this Port field belongs here or it depends on which type of backend is used?

Author: I purposefully don't want to wade into the named ports scenario since Service won't be a valid Backend type

Reviewer 8: endpointSlices has namedports (not IntOrStr but they have ports with names). We can followup on this later

Author: That's a good call; I added a unique name to a port type but that's it.

  // ExternalHostname specifies the configuration for an ExternalHostname backend. Only used if type is ExternalHostname.
  // Support: Extended
  // +optional

  ExternalHostname *ExternalHostnameBackend `json:"externalHostname,omitempty"`

  // EndpointSelector specifies the configuration for an EndpointSelector backend. Only used if type is EndpointSelector.
  // As defined in GEP-4731, creation of a `Backend` of type `EndpointSelector` should result in `Backend` controllers
  // creating the requisite `EndpointSelector` resource and setting ownerReferences appropriately.
  // TODO: Add link when GEP-4731 merges.
  // +optional

>   EndpointSelector *EndpointSelectorBackend `json:"endpointSelector,omitempty"`

Reviewer 8: are we willing to add services as a valid backend, in case a user wants to have Backend poiting directly to a clusterIP without caring about endpointSelector running on the cluster? Maybe I didn't reached it yet and endpointselector can be a pre-existing endpointSlices resource, in that case feel free to ignore me

Author: No backend MUST NOT send to a cluster IP because that will go through another round of load balancing. Endpoint selector is the answer for the backwards compat case

type BackendControllerStatus struct {
  // Name is a domain/path string that indicates the name of the controller that manages the
  // Backend. Name corresponds to the GatewayClass controllerName field when the
  // controller will manage parents of type "Gateway". Otherwise, the name is implementation-specific.
  //
  // Example: "example.net/import-controller".
  //
  // The format of this field is DOMAIN "/" PATH, where DOMAIN and PATH are valid Kubernetes
  // names (https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names).
  //
  // A controller MUST populate this field when writing status and ensure that entries to status
  // populated with their controller name are removed when they are no longer necessary.
  //
  // +required
  Name ControllerName `json:"name"`
  // For Kubernetes API conventions, see:
  // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
  // conditions represent the current state of the Backend resource.
  // Each condition has a unique type and reflects the status of a specific aspect of the resource.
  //
  // Standard condition types include:

>   // - "Available": the resource is fully functional

Reviewer 5: ok, this is maybe related to my question here https://github.com/kubernetes-sigs/gateway-api/pull/4488/changes#r3212983882 , is `fully functional` here meaning the producer/consumer config is validated and working? or is "fully functional on the consumer, but it may not work as expected if the configuration is wrong"

Author: The latter; we don't make guarantees that traffic to the endpoints will work

  // - "Progressing": the resource is being created or updated
  // - "Degraded": the resource failed to reach or maintain its desired state
  //
  // The status of each condition is one of True, False, or Unknown.
  // +listType=map
  // +listMapKey=type
  // +optional
  Conditions []metav1.Condition `json:"conditions,omitempty"`
}
```

### ExternalHostname Backend Configuration

```go
type ExternalHostnameBackend struct {
  // Hostname specifies the destination address used to reach this hostname.
  // IP addresses are not allowed in this field (enforced by validation on the type).

>   // TODO: Should we add additional validation to prevent *.cluster.local hostnames?

Reviewer 8: yes! but also, there are edge cases IIRC where a user can change the service suffix to something else.

In that case, we should probably add a MUST somewhere where the controller must reject attempts to use local cluster names, and names that are resolved to localhost or even the servicesCIDR or podCIDR

Author: > In that case, we should probably add a MUST somewhere where the controller must reject attempts to use local cluster names, and names that are resolved to localhost or even the servicesCIDR or podCIDR

Local cluster names based on a trust domain the controller is aware of in its service registry makes sense to me. I'm not so sure about checking resolution because that's going to result in the control plane making a ton of calls for every Backend

Author: I added validation and some language in line with what you said above for custom trust domains

  Hostname PreciseHostname `json:"hostname"`
}
```

### Protocol and Extension Support

```go
// +kubebuilder:validation:Enum=MCP
type BackendProtocol string

const (

>   BackendProtocolMCP   BackendProtocol = "MCP"

Reviewer 8: I can see some comments about it above. While we are adding the cases for MCP, I would expect that once we extend Backend usage for things like retry, timeout or session persistence we declare the backend something else like HTTP or TCP that will drive the behavior of the controller to accept or reject the configuration (eg.: maybe session persistence on a TCP backend does not make sense, but it does make sense for MCP or HTTP)

Author: 🤔 so the protocol has less to do with the gateway processing one protocol or another (the frontend pass will do that( and more to do with the sorts of properties (e.g http retries, session persistence, etc.) that the backend is configuring? I really like that actually

Author: Also need h2c (for servers that don't allow upgrade), http 1.1 (for the downgrade case).

2. **RBAC and Admission Control as Primary Security Control**
   - Application developers are the persona target by the `Backend` resource

>      - Implementations SHOULD provide control-plane guardrails (for example, an allow-list of permitted egress domains)

Reviewer 8: this may resolve my concern with the CEL validation above about cluster.local, and we must for sure test cases where localhost and others are rejected as a default by the implementation

Author: Yes agreed
